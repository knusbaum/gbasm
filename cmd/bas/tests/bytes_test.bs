package main

function main
	// Do the function prologue.
	prologue

	// HELLO WORLD\n\0
	bytes hello 13
	mov [hello+0] 0x48
	mov [hello+1] 0x45
	mov [hello+2] 0x4C
	mov [hello+3] 0x4C
	mov [hello+4] 0x4F
	mov [hello+5] 0x20
	mov [hello+6] 0x57
	mov [hello+7] 0x4F
	mov [hello+8] 0x52
	mov [hello+9] 0x4C
	mov [hello+10] 0x44
	mov [hello+11] 0x0A
	mov [hello+12] 0x00

	acquire rdi
	mov rdi hello
	call puts
	release rdi
	forget hello

	// GOODBYE WORLD\n\0
	bytes goodbye 15
	local ptr 64

	mov ptr goodbye
	mov [ptr+0] 0x47
	mov [ptr+1] 0x4F
	mov [ptr+2] 0x4F
	mov [ptr+3] 0x44
	mov [ptr+4] 0x42
	mov [ptr+5] 0x59
	mov [ptr+6] 0x45
	mov [ptr+7] 0x20
	mov [ptr+8] 0x57
	mov [ptr+9] 0x4F
	mov [ptr+10] 0x52
	mov [ptr+11] 0x4C
	mov [ptr+12] 0x44
	mov [ptr+13] 0x0A
	mov [ptr+14] 0x00

	acquire rdi
	mov rdi goodbye
	call puts
	release rdi
	forget goodbye

	bytes helloagain 13
	acquire rdi
	mov rdi helloagain
	call fillWithHelloWorld
	mov rdi helloagain
	call puts	
	release rdi
	
	epilogue
	xor rax rax
	ret
					
function fillWithHelloWorld
	argi ptr 0

	mov [ptr+0] 0x48
	mov [ptr+1] 0x45
	mov [ptr+2] 0x4C
	mov [ptr+3] 0x4C
	mov [ptr+4] 0x4F
	mov [ptr+5] 0x20
	mov [ptr+6] 0x57
	mov [ptr+7] 0x4F
	mov [ptr+8] 0x52
	mov [ptr+9] 0x4C
	mov [ptr+10] 0x44
	mov [ptr+11] 0x0A
	mov [ptr+12] 0x00	
