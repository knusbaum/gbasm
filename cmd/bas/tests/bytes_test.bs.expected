HELLO WORLD
GOODBYE WORLD
HELLO WORLD
